samples: net: sockets: big_http_download: Fix DHCP configuration

rlubos opened the pull request:
This PR fixes the `big_http_download` sample, which is currently broken due to DHCP behaviour. Changes in this PR:
* Update the DHCPv4 implementation, to prevent it from clearing the gateway configuration on initialization, therefore breaking the static IP configuration in case no DHCP server is available.
* Add send/recv timeout to the socket in the `big_http_download` sample to prevent it from stalling when the IP address on the interface changes.

rlubos commented:
@jukkar I was not sure if this should be rather fixed at the DHCP level (i.e. clearing the GW address) or the sample level, but I didn't find a good way to address the issue in that area (i. e. after successfull DHCP the IP address will likely change anyway, so there's little use to configure address statically first). Therefore I figured this should be rather addressed in the sample (configuring address both statically and with DHCP doesn't sound right in the first place to me).

pfalcon commented on samples/net/sockets/big_http_download/prj.conf:
```diff
@@ -27,9 +27,6 @@ CONFIG_NET_CONFIG_NEED_IPV4=y
 CONFIG_NET_CONFIG_PEER_IPV4_ADDR="192.0.2.2"
 CONFIG_NET_CONFIG_MY_IPV4_GW="192.0.2.2"
-# DHCP configuration. Until DHCP address is assigned,
-# static configuration above is used instead.
```
This comment effectively sets the requirement for networking stack configuration, and this sample effectively verifies this requirement. And the current state that we have a regression in the behavior. So, please don't remove this comment, but instead investigate what/when broke it. (I can help with that too, as it was in my backlog for some time.)

pfalcon reviewed (changes requested):
> Current IP configuration of the big_http_download sample, where we
condfigure IP addresses both statically and with DHCP does not work.

I noticed that too, and it actually broken since the 2.5 release. I was too shy to submit a ticket for it, because I wanted to investigate it myself (but keep being sidetracked and forget about it).

> but I didn't find a good way to address the issue in that area (i. e. after successfull DHCP the IP address will likely change anyway, so there's little use to configure address statically first).

It makes a lot of sense to configure static address and enable DHCP: this way, if you have DHCP infrastructure around (i.e., a network router which creates and oversees a whole local network), it will follow configuration of that network (e.g. IP address assignment). And if you don't have DHCP infrastructure around (i.e. you don't have a network router, but only have network "hosts" connected among themselves, in basic case, your laptop connected directly to a device) - you can still access the device by the static address.

This is how most of embedded Linux devices work. I recently had a case of debugging my embedded Linux NAS and greatly enjoyed this capability of being able to to see that device works and access it via network, instead of soldering JTAT cables, which would likely be the fate otherwise.

Zephyr is intended to also work like that, and I would consider that a serious regression. And instead of debugging/solving this issue, this patch only masks it. So, let me be explicitly -1 on it, and suggest to make a bisect to find what broke the correct behavior and look into fixing it.

rlubos commented:
> So, let me be explicitly -1 on it, and suggest to make a bisect to find what broke the correct behavior and look into fixing it.

The problem I faced (no DHCP server) was introduced by this change specifically:
https://github.com/zephyrproject-rtos/zephyr/commit/d44b4252b7fbfe8e02c807d2a6b3baf66876d8e2 (which came as a part of a larger rework https://github.com/zephyrproject-rtos/zephyr/pull/30995). The exact problem is that GW address is cleared on an interface on DHCP startup, so it's no longer possible to send packets to the outside network (we can discuss whether it's correct or not).

But, even with a DHCP server around (I've tested this scenario as well), I'm still not convinced if setting a static address first is a right approach. In my case, the sample established TCP connection, downloaded some data, and then after DHCP kicked in (it starts after a random backoff now), my device got assigned a new IP address, which became a problem. No more packets from the previously established connection could reach the device, so the open socket just kept hanging on the `recv()` call waiting for data that'd never come. 

I'm not sure what Linux does in this case, but maybe a correct way would be to invalidate the socket if the IP address it was bound to was removed from the system? Another point is whether the statically configured IP address should be removed in that case, probably not.

jukkar commented:
Having both static and dynamic (DHCP) address set to the interface can lead to these weird connectivity issues. Most likely the addresses are allocated from different subnets and we are not able to use the static addresses after the DHCP has managed to allocate an address. Then we are left with system that might not be able to communicate anywhere without first closing the sockets.

pfalcon commented:
@rlubos :

> The problem I faced (no DHCP server) was introduced by this change specifically:
d44b425 (which came as a part of a larger rework #30995). The exact problem is that GW address is cleared on an interface on DHCP startup,

Thanks for bisecting and finding the culprit commit. So, apparently clearing should happen not on "DHCP startup", but only when a valid IP address (and other config) was received/confirmed over DHCP.

> But, even with a DHCP server around (I've tested this scenario as well), I'm still not convinced if setting a static address first is a right approach.

I'm not sure the word "right" is the right one to use here. It should be *possible* to have both static IP config and DHCP enabled, and it should work as described above (and as Linux solutions work) - static IP config should be used until DHCP config is fully received and confirmed, and switchover should happen in as atomic way as realistically possible. What we have now is effectively a race condition between static config and DHCP, when just enabling DHCP leads to conflict with static config, even if DHCP won't really be used (i.e. when there's no DHCP server around, and DHCP requests sent from a Zephyr device won't be responded to). 

> In my case, the sample established TCP connection, downloaded some data, and then after DHCP kicked in (it starts after a random backoff now), my device got assigned a new IP address, which became a problem. 

And in my case (I usually run with direct device-host connection, i.e. without DHCP), the app downloaded some data, then DHCP kicked in and overthrown the proper config, and everything hanged. And the only solution is to disable DHCP in the config. That's why I say we have been having a regression for some time now, and why I say this patch just masks the issue instead of properly resolving it. (Again, this sample specifically had this configuration to ensure that static IP + DHCP works in an expected way - failing us to have a proper unit test for that.)

> No more packets from the previously established connection could reach the device, so the open socket just kept hanging on the recv() call waiting for data that'd never come.

That's orthogonal issue. It's quite well-known that one can't write real-world robust applications without timeouts for I/O (in particular, networking) operations. This sample was intended to be a simple one, and was written before we had timeout support for recv/send. We now have it (TODO: check that we actually have support for both by now), so if you'd like to add timeout for recv() here, I'm +1.

Note that IP may change under app's feet at any time. For example, DHCP has lease time, when it ends, a device should query again and may get a different IP address, and lease time may be 5min or so. That's again why I say that this patch just masks the issue, and that instead we should value this sample's config as something which exposes the issue and allows us to manage it (hopefully even improve over time).

Finally, we have had https://github.com/zephyrproject-rtos/zephyr/issues/7630 for quite some time, raising attention to these issues. That's why I'm keen to at least preserve/recover previous state of affairs, and not mask/deteriorate it.

> I'm not sure what Linux does in this case, but maybe a correct way would be to invalidate the socket if the IP address it was bound to was removed from the system?

These ideas is what's #7630 dedicated to. But those are outside BSD Sockets API, and the baseline is that interface vs sockets layers are separate in the network stack, and each works independently, and there's a generic timeout mechanism available for apps to catch *any* connectivity issues on the socket level - timeout.

> Another point is whether the statically configured IP address should be removed in that case, probably not.

I see no ambiguity here (taking Linux experience as the baseline) - if DHCP is enabled on/for interface, it overrides interface's IP address - once the address is successfully received. From Linux PoV, if I'd like to have both static and DHCP address on the same interface, I'd need to create a virtual sub-interface (e.g. eth0.1), and have static IP on eth0 and run DHCP on eth0.1. (Other trickier option may be to settle that DHCP overrides if's *first* IP address, so way to have both DHCP and static is to configure the latter as 2nd).

pfalcon commented:
@jukkar 

> Having both static and dynamic (DHCP) address set to the interface can lead to these weird connectivity issues.

The talk is more about co-existence of static and DHCP configuration on the interface, where static is the default/fallback, and DHCP overrides it, once successfully received - without race conditions.

> Most likely the addresses are allocated from different subnets and we are not able to use the static addresses after the DHCP has managed to allocate an address. Then we are left with system that might not be able to communicate anywhere without first closing the sockets.

That's realistic situation anyway (any network connection may break/stall for many reasons, including those beyond IP addresses being changed behind app's back), so an app should be ready to deal with those (standard generic means for that is having timeouts on I/O operations, and when those fail, retry the entire connection attempt).

jukkar commented:
> The talk is more about co-existence of static and DHCP configuration on the interface, where static is the default/fallback, and DHCP overrides it, once successfully received - without race conditions.

I was just wondering the way we have this working atm, does it make actually sense to have this kind of "fallback" as it can/is causing issues when running in real life environment.

pfalcon commented:
@jukkar 

> I was just wondering the way we have this working atm, does it make actually sense to have this kind of "fallback" as it can/is causing issues when running in real life environment.

We have had that fallback all the time, and it didn't cause any issues in real life environment, I specifically verified and tested it with this sample, that's why these options are enabled in it. Until it regressed before 2.5 (apparently with the 
commit https://github.com/zephyrproject-rtos/zephyr/commit/d44b4252b7fbfe8e02c807d2a6b3baf66876d8e2, as pointed by @rlubos). And I think we should recover that behavior, rather than mask the issue.

rlubos commented:
Thank you for all the feedback. Please correct me if I misunderstood something, but the way to go for now is to restore the original behavior by:
* Sorting out the DHCPv4 client behaviour, which clears the gateway configuration on init (which should fix the scenario where no DHCP server is available). 
* Configuring timeout on a socket with `SO_RCVTIMEO`/`SO_SNDTIMEO` so that the sample no longer stalls after the IP address configured on an interface changes.

I'll try to propose something soon(ish).

rlubos committed:
samples: net: sockets: big_http_download: Set RX/TX timeout on a socket

Configure a timeout for send/recv operations so that the sample can
recover in case the connection stalls (for instance due to IP address
change on an interface).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: dhcpv4: Prevent DHCP client from clearing the gateway

In case both, static IP address configuration and DHCP were used, and no
DHCP server was avaliable in the network, clearing the gateway address
rendered the network interface unusable as it's gateway configuration
was cleared.

Prevent this by removing the gateway clearing during the DHCP
inititalization. If the DHCP server is available in the network, the
gateway address will be overwriten after receiving the DHCP OFFER
message or cleared if there's no Router option is avaiable in the
DHCP OFFER message.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@jukkar @pfalcon I've pushed new commits, as discussed above, and updated the PR description.

As for the DHCP, the gateway IP address is updated when Router option is detected in a DHCP message. In my setup, the option was present in both, DHCP OFFER and DHCP ACK messages. however I didn't find any information in the RFC that the option is mandatory (and likely isn't since it's listed in the "Vendor extensions" section). Therefore I've kept the gateway clearing but only in a case, when there is no Router option in the DHCP OFFER message sent by the DHCP server.

pfalcon commented:
@rlubos:

>  *   Sorting out the DHCPv4 client behaviour, which clears the gateway configuration on init (which should fix the scenario where no DHCP server is available).
> *    Configuring timeout on a socket with SO_RCVTIMEO/SO_SNDTIMEO so that the sample no longer stalls after the IP address configured on an interface changes.

Sorry for the delay with response, and yes, that's what I meant. These 2 things are of course orthogonal and could be done separately (e.g. if DHCP issue turns out to take more time), but I appreciate give a try to both. Testing on my side...

pfalcon reviewed (approved):
Tested on following configs:

* qemu_x86 w/o DHCP
* frdm_k64f w/o DHCP (direct connection to a laptop)
* frdm_k64f with DHCP (connected to an OpenWRT router)

Works as expected in all cases (i.e. in case of no DHCP, proceeds with download undisturbed, assuming static IP config is correct; in case of DHCP, there's a few retries until DHCP config is received).

Thanks for taking care of this!
